pandas/io/formats/format.py

```diff
@@ -1622,17 +1622,31 @@ def __init__(
     def _format_strings(self) -> list[str]:
         """we by definition have DO NOT have a TZ"""
         values = self.values
+        values_shape = values.shape
+        values = values.ravel() if values.ndim > 1 else values
 
         if not isinstance(values, DatetimeIndex):
             values = DatetimeIndex(values)
 
         if self.formatter is not None and callable(self.formatter):
-            return [self.formatter(x) for x in values]
+            values = [self.formatter(x) for x in values]
+        else:
+            values = values._data._format_native_types(
```

Comment:
The relevant methods should already support 2D, so i think we only need to fix/change the self.formatter line above?

Reply:
Ok Will change the self.formatter

Reply:
might be worth looking at np.frompyfunc, ive been looking at using this as a replacement for ravel_compat to handle 2D cases

Reply:
Hi I am in loss to understand this.  After run steps inside Pandas code, step go into this function
In the pandas\core\arrays\datetimes.py,
```
@dtl.ravel_compat
def _format_native_types(
    self, *, na_rep="NaT", date_format=None, **kwargs
) -> npt.NDArray[np.object_]:
    from pandas.io.formats.format import get_format_datetime64_from_values

    fmt = get_format_datetime64_from_values(self, date_format)

    return tslib.format_array_from_datetime(
        self.asi8, tz=self.tz, format=fmt, na_rep=na_rep
    )
```
Are you mean to replace this function from ravel_compat  to np.frompyfunc in this file? Hope can give some guidances. Thank. On my side, I try best to figure this out.

Reply:
my thought was that L1632 might end up looking something like `values = np.frompyfunc(self.formatter, 1, 1)(values)`

Reply:
Hi, I have try to change code to 
```
def _format_strings(self) -> list[str]:
        """we by definition have DO NOT have a TZ"""

        values = self.values

        if not isinstance(values, DatetimeIndex):
            values = DatetimeIndex(values)

        if self.formatter is not None and callable(self.formatter):
            fmt_values = np.frompyfunc(self.formatter, 1, 1)(values)
        else:
            fmt_values = values._data._format_native_types(
                na_rep=self.nat_rep, date_format=self.date_format
            )

        nested_formatter = GenericArrayFormatter(fmt_values)
        fmt_values = nested_formatter.get_result()
        return list(fmt_values)
```
I test this.
```
x = date_range("2018-01-01", periods=24, freq="H").to_numpy()

def format_func(t):
    return t.strftime("%H-%m")

formatter = fmt.Datetime64Formatter(x.reshape((4, 2, 3)), formatter=format_func)
result = formatter.get_result()
```

The error looks like this
```
  from pandas import (
Traceback (most recent call last):
  File "D:\kcw\development\testing.py", line 73, in <module>
    result = formatter.get_result()
  File "d:\kcw\pandas\pandas\io\formats\format.py", line 1349, in get_result
    fmt_values = self._format_strings()
  File "d:\kcw\pandas\pandas\io\formats\format.py", line 1631, in _format_strings
    fmt_values = np.frompyfunc(self.formatter, 1, 1)(values)
  File "d:\kcw\pandas\pandas\core\indexes\base.py", line 949, in __array_ufunc__
    result = getattr(ufunc, method)(*new_inputs, **kwargs)
  File "d:\kcw\pandas\pandas\core\arrays\datetimelike.py", line 1793, in __array_ufunc__
    return super().__array_ufunc__(ufunc, method, *inputs, **kwargs)
  File "d:\kcw\pandas\pandas\core\arrays\base.py", line 1648, in __array_ufunc__
    return arraylike.default_array_ufunc(self, ufunc, method, *inputs, **kwargs)  File "d:\kcw\pandas\pandas\core\arraylike.py", line 490, in default_array_ufunc
    return getattr(ufunc, method)(*new_inputs, **kwargs)
  File "D:\kcw\development\testing.py", line 70, in format_func
    return t.strftime("%H-%m")
AttributeError: 'int' object has no attribute 'strftime'
```
But I encounter this problem. In this line `fmt_values = np.frompyfunc(self.formatter, 1, 1)(values)` . The np.frompyfunc function converts the values <class 'pandas.core.indexes.datetimes.DatetimeIndex'> to integer (ns).

Reply:
Hi. 
1. For this class `Datetime64Formatter`,  
```
x = date_range("2018-01-01", periods=10, freq="H").to_numpy()
values = DatetimeIndex(x)
fmt_values = np.frompyfunc(self.formatter, 1, 1)(values)
def format_func(t):
return t.strftime("%H-%m")

The np.frompyfunc will convert `2018-01-01` to ns integer value which